### PR TITLE
feature(api): Allow push_out arg in the dispense command

### PR DESCRIPTION
**Pull request by TamarZanzouri**
# Overview

closes https://opentrons.atlassian.net/browse/RSS-308.
allow push_out arg in the dispense command for PAPI and PE.
update all schema 2 pipette FLEX configuration with default push out volume. 

# Test Plan

- make sure you are able to use python protocols with a push_out arg in the dispense command.
- make sure you are able to POST http dispense command with a push_out arg.
- make sure supplying a negative value raises.

# Changelog

- Added an optional `push_out` arg to the dispense commands in PAPI and in PE. 
- Added `default_push_out_volume` to PipetteDict
- Added optional `push_out` arg to hardware controller - not implemented hardware layer - will be done in a separate ticket. 
- Updated all pipette schema v2 definitions to include `default_push_out_volume`.

# Review requests

- make sure the numbers are correct (numbers from andy in the ticket)
- changes make sense?
- am I missing something?

# Risk assessment

low. currently does not do anything since the hardware layer is not implemented. adds a skeleton for the `push_out` arg and the `defualt_push_out_volume` and raises errors for analysis and runs.

### CODE REVIEW

**Review by Laura-Danielle — COMMENTED** [review]
It's a bit annoying, but imo we should have the defaultPushOutVolume be an optional pydantic field.

It should default to 0.

File: api/src/opentrons/hardware_control/api.py [hunk 1]
```diff
@@ -986,6 +986,7 @@ async def aspirate(
     async def dispense(
         self,
         mount: top_types.Mount,
+        push_out: Optional[float] = None,
         volume: Optional[float] = None,
         rate: float = 1.0,
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/api.py:990]: Could we make push_out the last parameter here?

File: api/src/opentrons/hardware_control/protocols/liquid_handler.py [hunk 2]
```diff
@@ -66,6 +66,7 @@ async def aspirate(
     async def dispense(
         self,
         mount: Mount,
+        push_out: Optional[float],
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/protocols/liquid_handler.py]: Same here, can we put this parameter at the end? Also it's nice to default to `None` if we're making an optional parameter.

File: api/src/opentrons/protocol_api/core/engine/instrument.py [hunk 3]
```diff
@@ -155,6 +156,7 @@ def dispense(
             rate: Not used in this core.
             flow_rate: The flow rate in µL/s to dispense at.
             in_place: whether this is a in-place command.
+            push_out: The amount to push the plunger bellow bottom position.
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_api/core/engine/instrument.py]: `below`

File: api/src/opentrons/protocol_engine/commands/load_labware.py [hunk 4]
```diff
@@ -90,6 +93,18 @@ def __init__(
 
     async def execute(self, params: LoadLabwareParams) -> LoadLabwareResult:
         """Load definition and calibration data necessary for a labware."""
+        # TODO (tz, 8-15-2023): extend column validation to column 1 when working
+        # on https://opentrons.atlassian.net/browse/RSS-258 and completing
+        # https://opentrons.atlassian.net/browse/RSS-255
+        if (
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_engine/commands/load_labware.py]: Accidental change here?

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 5]
```diff
@@ -88,15 +89,22 @@ async def dispense_in_place(
...
+        push_out: Optional[float],
     ) -> float:
         """Dispense liquid without moving the pipette."""
         hw_pipette = self._state_view.pipettes.get_hardware_pipette(
             pipette_id=pipette_id,
             attached_pipettes=self._hardware_api.attached_instruments,
         )
 
+        if push_out and (
+            push_out < 0 or push_out > hw_pipette.config["plunger_positions"].bottom
```
Comment (Laura-Danielle) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]: I'm very confused. Push out is a volume, so why are we comparing it to the plunger position which is a mm value?

**Review by Laura-Danielle — CHANGES_REQUESTED** [review]
- We don't need plunger positions to be in the protocol engine for the push_out feature
- You have some labware changes that snuck into this branch related to trash validation -- I think we need to move it
- Have some other smaller requests as well as the comment above this change request comment.

File: api/src/opentrons/protocol_engine/resources/pipette_data_provider.py [hunk 6]
```diff
@@ -31,6 +31,8 @@ class LoadedStaticPipetteData:
         float, pipette_definition.SupportedTipsDefinition
     ]
     nominal_tip_overlap: Dict[str, float]
+    default_push_out_volume: Optional[float]
+    plunger_positions_for_liquid_class: pipette_definition.PlungerPositions
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/resources/pipette_data_provider.py]: can remove

File: api/src/opentrons/protocol_engine/execution/pipetting.py (hunk 5 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_engine/execution/pipetting.py]: create a ticket for a floating method to get this value (bottom poistion + default blow out converted to ul)

**Review by Laura-Danielle — CHANGES_REQUESTED** [review]
I think you still have some accidental changes in this branch that needs to be cleaned up. I'd like to rebase on top of this, but I'll hold off for now.

File: shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/2_2.json [hunk 7]
```diff
@@ -97,7 +97,8 @@
           ]
         }
       },
-      "defaultBlowoutVolume": 127.235
+      "defaultBlowoutVolume": 127.235,
+      "defaultPushOutVolume": 20
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/2_2.json]: I think we want to leave all GEN2 pipettes alone, so this should be `0` I believe

File: shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/2_1.json [hunk 8]
```diff
@@ -97,7 +97,8 @@
           ]
         }
       },
-      "defaultBlowoutVolume": 127.235
+      "defaultBlowoutVolume": 127.235,
+      "defaultPushOutVolume": 20
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/2_1.json]: same here

File: shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/2_0.json [hunk 9]
```diff
@@ -95,7 +95,8 @@
           ]
         }
       },
-      "defaultBlowoutVolume": 127.235
+      "defaultBlowoutVolume": 127.235,
+      "defaultPushOutVolume": 20
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/2_0.json]: same here

File: shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_5.json [hunk 10]
```diff
@@ -30,7 +30,8 @@
       "dispense": {
         "default": { "1": [[1000.0, 0.0, 61.3275]] }
       },
-      "defaultBlowoutVolume": 127.235
+      "defaultBlowoutVolume": 127.235,
+      "defaultPushOutVolume": 20
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_5.json]: same here

File: shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_4.json [hunk 11]
```diff
@@ -32,7 +32,8 @@
           "1": [[1000.0, 0.0, 61.3275]]
         }
       },
-      "defaultBlowoutVolume": 127.235
+      "defaultBlowoutVolume": 127.235,
+      "defaultPushOutVolume": 20
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_4.json]: same here

File: shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_3.json [hunk 12]
```diff
@@ -32,7 +32,8 @@
           "1": [[1000.0, 0.0, 61.3275]]
         }
       },
-      "defaultBlowoutVolume": 127.235
+      "defaultBlowoutVolume": 127.235,
+      "defaultPushOutVolume": 20
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_3.json]: same here

File: shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_0.json [hunk 13]
```diff
@@ -30,7 +30,8 @@
       "dispense": {
         "default": { "1": [[1000.0, 0.0, 61.3275]] }
       },
-      "defaultBlowoutVolume": 127.235
+      "defaultBlowoutVolume": 127.235,
+      "defaultPushOutVolume": 20
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/single_channel/p1000/default/1_0.json]: same here

File: shared-data/pipette/definitions/2/liquid/eight_channel/p50/default/1_4.json [hunk 14]
```diff
@@ -36,7 +36,8 @@
           "2": [[50, 0, 3.06368702]]
         }
       },
-      "defaultBlowoutVolume": 4.712
+      "defaultBlowoutVolume": 4.712,
+      "defaultPushOutVolume": 2
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/eight_channel/p50/default/1_4.json:40]: same here

File: shared-data/pipette/definitions/2/liquid/eight_channel/p50/default/1_3.json [hunk 15]
```diff
@@ -36,7 +36,8 @@
           "2": [[50, 0, 3.06368702]]
         }
       },
-      "defaultBlowoutVolume": 4.712
+      "defaultBlowoutVolume": 4.712,
+      "defaultPushOutVolume": 2
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/eight_channel/p50/default/1_3.json:40]: same here

File: shared-data/pipette/definitions/2/liquid/eight_channel/p50/default/1_0.json [hunk 16]
```diff
@@ -36,7 +36,8 @@
           "2": [[50, 0, 3.06368702]]
         }
       },
-      "defaultBlowoutVolume": 4.712
+      "defaultBlowoutVolume": 4.712,
+      "defaultPushOutVolume": 2
```
Comment (andySigler) [inline shared-data/pipette/definitions/2/liquid/eight_channel/p50/default/1_0.json:40]: same here

**Review by andySigler — APPROVED** [review]
shared-data changes look good to me, other than the older-generation pipettes getting a few non-zero push-out volumes but those have no affect on an OT2 anyways so looks good to me.

File: api/src/opentrons/protocol_engine/commands/dispense_in_place.py [hunk 17]
```diff
@@ -23,7 +23,9 @@
 class DispenseInPlaceParams(PipetteIdMixin, VolumeMixin, FlowRateMixin):
     """Payload required to dispense in place."""
 
-    pass
+    pushOut: Optional[float] = Field(
+        None, description="perform a small blow out for accurate dispensing"
```
Comment (sfoster1) [inline api/src/opentrons/protocol_engine/commands/dispense_in_place.py]:
```suggestion
        None, description="push the plunger a small amount farther than necessary for accurate low-volume dispensing"
```
i guess? @andySigler what do you think

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Agree with @andySigler about leaving gen2 and gen1 alone, let's keep those all 0. other than that looks great, thank you!

**Review by Laura-Danielle — APPROVED** [review]
nice, lgtm!

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

File: shared-data/command/schemas/7.json [hunk 18]
```diff
@@ -460,6 +460,11 @@
           "title": "Pipetteid",
           "description": "Identifier of pipette to use for liquid handling.",
           "type": "string"
+        },
+        "pushOut": {
+          "title": "Pushout",
+          "description": "push the plunger a small amount farther than necessary for accurate low-volume dispensing",
+          "type": "number"
```
Comment (SyntaxColoring) [inline shared-data/command/schemas/7.json:467]: I think we also need this new argument in `JsonTranslator`. Currently, it's getting silently discarded from `dispense` and `dispenseInPlace` commands in JSON protocols.

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 19]
```diff
@@ -88,15 +93,22 @@ async def dispense_in_place(
...
         flow_rate: float,
+        push_out: Optional[float],
     ) -> float:
         """Dispense liquid without moving the pipette."""
         hw_pipette = self._state_view.pipettes.get_hardware_pipette(
             pipette_id=pipette_id,
             attached_pipettes=self._hardware_api.attached_instruments,
         )
-
+        # TODO (tz, 8-23-23): add a check for push_out not larger that the max volume allowed when working on this https://opentrons.atlassian.net/browse/RSS-329
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/execution/pipetting.py:103]: You might already have this in mind, but I think we *also* need to add a check for `push_out` only being present when we're not expecting any liquid to be left in the tip. (If my memory from in-person conversations is correct.)

**Review by SyntaxColoring — COMMENTED** [review]
Tested Python and a JSON protocol. Mostly looks good to me, but a couple of notes/questions.

File: shared-data/command/schemas/7.json (hunk 18 above)
Comment (TamarZanzouri) [reply shared-data/command/schemas/7.json:467]: Thank you Max! Ill add it!

Comment (TamarZanzouri) [reply shared-data/command/schemas/7.json:467]: added a fix for this in the embedded pr. https://github.com/Opentrons/opentrons/pull/13422

File: api/src/opentrons/protocol_engine/execution/pipetting.py (hunk 19 above)
Comment (sfoster1) [reply api/src/opentrons/protocol_engine/execution/pipetting.py:103]: Agree. Will do in a followup

**Pull request merged**